### ISSUE TITLE
Add a `photon` unit?

### ISSUE THREAD

**Issue by kjohnsen** [brian-team/brian2#1211]
Hello there,

I'm working on modeling optogenetics and find myself needing to work with photons, but Brian2 doesn't have a photon unit. Maybe it's not a good idea to add every unit people suggest, but in my case it would be nice. I know I can work around by simply treating photons as a constant. 

As an alternative, if lots of people have cases like mine, has a system of users being able to define their own units been considered?

**Comment by thesamovar** [comment]
We considered this but there are a very large number of 'dimensionless' units like this we could add (e.g. spikes), and allowing for that would make the implementation much less efficient, so we decided against. We could potentially allow you to define new names for units (so photon would just be an alias for 1). It wouldn't catch unit errors since it would just internally be a dimensionless constant, but might allow for a slightly nicer way of annotating code. Would that be useful? Thoughts @mstimberg?

**Comment by mstimberg** [comment]
In general you can extend the unit system (e.g. you could write something like `photon = Unit(1, name="photon", dispname="photon")`), but dimensionless values are handled in a particular way and you'd never get to see that unit name.

Could you give some more details where you'd like to use this? Is this about units in equations, say you'd like to be able to write `photon_rate : photon/second`?

A general point that we should make clearer in the documentation (and maybe even in the syntax) is that in most places in Brian we are storing *dimensions* and not *units*. For example, whether you define a variable as having units of `volt` or `ohm*amp` in an equation does not make a difference, since both have the same dimensions. For dimensionless variables, we only store the information that it is dimensionless and nothing else, limiting our options here.

We could potentially store additionally store whether a variable is *angular* (which would also get rid of the issue that you sometimes see dimensionless values been shown as having "units of radian"). E.g. it does not make sense to calculate `cos(x)` if `x` is dimensionless but not angular (e.g. photons, spikes, ...). But this is not what you asked about :)

**Comment by kjohnsen** [comment]
Thanks! I suppose the corresponding unit for photons would be moles then, right, since we're counting them? Maybe there are people out there who would really appreciate the option of unit-checking on a dimension they've defined, but it sounds like it's probably not worth worrying about.
This is how I would use photon if it were there: `phi = Irr / Ephoton : photon/second/meter2`, where `phi` is the photon flux calculated from the irradiance coming from the light source.

**Comment by mstimberg** [comment]
Right, I did not actually think about the "quantity" dimension, I agree that photon would rather be like moles instead of dimensionless. The main problem still remains, though: Brian's unit system does not make a difference between different units with the same dimensions. E.g. if you defined your own unit `photon` (which is possible), then `1e6*photon` and `1e6*mole` would still be stored in the exact same way, so you could not make a difference between the two (and the string output would rather be in `mole`, for example). Overhauling the unit system to store the units instead of just the dimensions would be a major task and I don't think we are going to tackle this anytime soon...
